x/reward/keeper/rules.go

```diff
@@ -0,0 +1,241 @@
+package keeper
+
+import (
+	"fmt"
+
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/provenance-io/provenance/x/reward/types"
+)
+
+func (k Keeper) ProcessTransactions(ctx sdk.Context) {
+	// Get all Active Reward Programs
+	rewardPrograms, err := k.GetAllActiveRewardPrograms(ctx)
+	if err != nil {
+		ctx.Logger().Error(err.Error())
+		return
+	}
+
+	// Grant shares for qualifying actions
+	for index := range rewardPrograms {
+		// Go through all the reward programs
+		actions, err := k.DetectQualifyingActions(ctx, &rewardPrograms[index])
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+			continue
+		}
+
+		// Record any results
+		err = k.RewardShares(ctx, &rewardPrograms[index], actions)
+		if err != nil {
+			ctx.Logger().Error(err.Error())
+		}
+	}
+}
+
+// DetectQualifyingActions takes in the RewardProgram and checks if any of the qualifying actions is found in the event history
+func (k Keeper) DetectQualifyingActions(ctx sdk.Context, program *types.RewardProgram) ([]types.EvaluationResult, error) {
+	ctx.Logger().Info(fmt.Sprintf("EvaluateRules for RewardProgram: %d", program.GetId()))
+	results := []types.EvaluationResult(nil)
+
+	// Check if any of the transactions are qualifying actions
+	for _, supportedAction := range program.GetQualifyingActions() {
+		// Get the appropriate RewardAction
+		// If it's not supported we skip it
+		action, err := supportedAction.GetRewardAction(ctx)
+		if err != nil {
+			ctx.Logger().Info(err.Error())
+			continue
+		}
+
+		// Build all the qualifying actions from the abci events
+		actions, err := k.FindQualifyingActions(ctx, action)
+		if err != nil {
+			return nil, err
+		}
+
+		// Process actions and get the results
+		actions = k.ProcessQualifyingActions(ctx, program, action, actions)
+		results = append(results, actions...)
+	}
+
+	return results, nil
+}
+
+func (k Keeper) ProcessQualifyingActions(ctx sdk.Context, program *types.RewardProgram, processor types.RewardAction, actions []types.EvaluationResult) []types.EvaluationResult {
+	successfulActions := []types.EvaluationResult(nil)
```

Comment:
This change would lead to an unallocated array. We currently have a slice, but we can allocate space upfront if we think slice appending is a bottleneck.

Reply:
`successfulActions := []types.EvaluationResult(nil)` casts `nil` into a `[]types.EvEvaluationResult`.
`var successfulActions []types.EvaluationResult` creates a variable of type `[]types.EvaluationResult` with a value of `nil`.

They're almost equivalent. The former requires a cast operation. It's also syntax that is rarely used for defining a slice variable.